cmd/gitlab.go

```diff
@@ -362,74 +138,84 @@ func (c *Client) pollPipelineJobs(p Project, pipelineID int, topics string, ref
+		return fmt.Errorf("could not read content of last pipeline %s:%s", gitlabProject.PathWithNamespace, ref)
 	}
 	if lastPipeline != nil {
-		runCount.WithLabelValues(gp.PathWithNamespace, topics, ref).Inc()
+		runCount.WithLabelValues(gitlabProject.PathWithNamespace, topics, ref).Inc()
 
 		if lastPipeline.Coverage != "" {
 			parsedCoverage, err := strconv.ParseFloat(lastPipeline.Coverage, 64)
 			if err != nil {
 				log.Warnf("Could not parse coverage string returned from GitLab API '%s' into Float64: %v", lastPipeline.Coverage, err)
 			}
-			coverage.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(parsedCoverage)
+			coverage.WithLabelValues(gitlabProject.PathWithNamespace, topics, ref).Set(parsedCoverage)
 		}
 
-		lastRunDuration.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(float64(lastPipeline.Duration))
-		lastRunID.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(float64(lastPipeline.ID))
+		lastRunDuration.WithLabelValues(gitlabProject.PathWithNamespace, topics, ref).Set(float64(lastPipeline.Duration))
+		lastRunID.WithLabelValues(gitlabProject.PathWithNamespace, topics, ref).Set(float64(lastPipeline.ID))
 
-		c.outputPipelineStatusMetric(lastPipeline.Status, p.ShouldOutputSparseStatusMetrics(cfg), gp.PathWithNamespace, topics, ref)
+		emitStatusMetric(
+			lastRunStatus,
+			[]string{gitlabProject.PathWithNamespace, topics, ref},
+			statusesList,
+			lastPipeline.Status,
+			outputSparseStatusMetrics,
+		)
 
-		if p.ShouldFetchPipelineJobMetrics(cfg) {
-			if err := c.pollPipelineJobs(p, lastPipeline.ID, topics, ref); err != nil {
+		if fetchPipelineJobMetrics {
+			if err := c.pollPipelineJobs(gitlabProject, lastPipeline.ID, topics, ref, outputSparseStatusMetrics); err != nil {
 				log.Errorf("Could not poll jobs for pipeline %d: %s", lastPipeline.ID, err.Error())
 			}
 		}
 
-		timeSinceLastRun.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(time.Since(*lastPipeline.CreatedAt).Round(time.Second).Seconds())
+		timeSinceLastRun.WithLabelValues(gitlabProject.PathWithNamespace, topics, ref).Set(time.Since(*lastPipeline.CreatedAt).Round(time.Second).Seconds())
 	}
 
 	return nil
 }
 
-func (c *Client) pollProjects(until <-chan bool) {
+func (c *Client) orchestratePolling(until <-chan bool, getRefsOnInit <-chan bool) {
 	go func() {
 		pollWildcardsEvery := time.NewTicker(time.Duration(cfg.ProjectsPollingIntervalSeconds) * time.Second)
 		pollRefsEvery := time.NewTicker(time.Duration(cfg.RefsPollingIntervalSeconds) * time.Second)
 		stopWorkers := make(chan struct{})
 		defer close(stopWorkers)
-		// first execution
+		// first execution, blocking call before entering orchestration loop to get the wildcards discovered
 		c.discoverWildcards()
-		c.pollWithWorkersUntil(stopWorkers)
+
 		for {
 			select {
 			case <-until:
 				log.Info("stopping projects polling...")
 				return
+			case todo := <-getRefsOnInit:
+				log.Debugf("executing init pipeline? %v", todo)
```

Comment:
I believe this should be within the if statement underneath. Also, a more pertinent message could be nice!

Reply:
agree on the message but the reason for it to be external is to debug what value is passed _before_ evaluating.

Corrected with
```
log.Debugf("should we poll the more recent refs from the last executed pipelines? %v", todo)
```

Reply:
oh my bad 🤦 sounds good 👍